### ISSUE TITLE
JREF and NULL objects - Doc and Code

### ISSUE THREAD

**Issue by ssardina** [SWI-Prolog/packages-jpl#20]
Hi @anionic 

While I am redoing the test so that we have no failure, almost there, I realized that the tests, API and doc for JREF is a bit out-dated.

I suspect that JREF was introduced later, and that objects were somehow deal with special Compund @ terms.

The doc says:

`here is no org.jpl7.JNull class: instead, use new JRef(null) to represent @(null) (which itself represents Java's null). If you don't know what this all means, don't worry: it only affects those writing hybrid Java+Prolog programs which call each other nestedly.`

This may be old, as the constructor for JREF will throw an exception if you give null:

`JPLException("a JRef cannot reference null (use JPL.newJRef() or JPL.JNULL)")`

We may want to discuss how we handle JREF and in particular NULL objects and their JPL JNULL counterpart. Since we do have a JNULL I added support in the `.object()` method of `Term`:

```
	/**
	 * The Object which this org.jpl7.JRef refers to, iff this Term is a JRef or just JPL.JNULL.
	 *
	 * @return the Object which this Term refers to, iff this Term is a JRef.
	 * @throws JPLException if this Term is not a JRef or not NULL term
	 */
	public Object object() { // overridden in JRef
		if (this == JPL.JNULL)
		    return null;
        else
            throw new JPLException("this term is not a JRef");
	}

```

Then, if the Term is itself a JREF of a non-null, then the object itself will be returned:

@anionic , are you happy with this? We need to update your doc and in fact add a whole section on `JREF` as there is almost nothing except for a note at the end which needs correction. It appears to me that you used to use `@` for JREF?

**Comment by anionic** [comment]
First, we need a plausible use case for org.jpl7.JRef

Users calling 3rd party Prolog code from Java will not need it.

Users calling 3rd party Java code from Prolog will not need it.

Only those writing hybrid Prolog+Java apps which call each other both ways *may* need it, or at least want it.

I removed it years ago, but was begged to restore it by one user who had found it convenient and didn't want to rewrite his/her code.

I don't want to confuse mainstream users by drawing attention to it in the docs.

In its present form it has been overtaken by the 7.4 re-implementation of JRefs from e.g. @('J#00000000000443566616') to `<jref>(0x2193410)` (blob)

As it is currently broken, and no-one has complained since this change, I am inclined to remove it (again), but first need to dig up that use case and demonstrate that it isn't essential.

**Comment by ssardina** [comment]
> First, we need a plausible use case for org.jpl7.JRef
> 

I am not following. You guys added it and is already there in JPL. And it is really cool feature.

> Users calling 3rd party Prolog code from Java will not need it.

correct.

> 
> Users calling 3rd party Java code from Prolog will not need it.

correct if you just call static methods right?



> Only those writing hybrid Prolog+Java apps which call each other both ways _may_ need it, or at least want it.
> 

correct.

> I removed it years ago, but was begged to restore it by one user who had found it convenient and didn't want to rewrite his/her code.
> 
> I don't want to confuse mainstream users by drawing attention to it in the docs.
> 
> In its present form it has been overtaken by the 7.4 re-implementation of JRefs from e.g. @('J#00000000000443566616') to `<jref>(0x2193410)` (blob)
> 
> As it is currently broken, and no-one has complained since this change, I am inclined to remove it (again), but first need to dig up that use case and demonstrate that it isn't essential.

Noooooooo please. Do not remove it! :-)

Indeed, I am another one who wants it because I can see the use. 

The whole point for me of JPL is to make Prolog accessible to "mainstream" programming. Allowing Prolog to call your Java objects is needed to avoid repetition of data (which is error prone and requires lots of sync). You may have data that is on the Java side and you cannot just move it to the Prolog side, it needs to remain in Java (for many reasons, including legacy). If you cannot access that data from Prolog one needs to duplicate it and keep it updated.

As a silly example, I may have an object in Java denoting a person with its personal details and want to access them from Prolog without duplicating the person there.

**Why do you say it is broken?**

**Comment by anionic** [comment]
Don't panic! You misunderstand. **Prolog-side** jrefs are central to JPL, a "cool feature" indeed:
```
?- jpl_new('java.util.Date', [], JRef), jpl_call(JRef, toString, [], S).
JRef = <jref>(00000137ECA1E840),
S = 'Sat Oct 13 12:52:36 BST 2018'.
```
(I maintain production code which depends on them and pays my salary) but I question whether we need a dedicated Java-side concrete syntax for them. Do you have a code snippet doing something reasonable which cannot be done without using org.jpl7.JRef? I think this would have to be Java code which incestuously creates Prolog goals which call back into Java ;-) and maybe this should instead be supported by extending org.jpl7.Atom to represent blobs of type jref.

**Comment by ssardina** [comment]
> Don't panic! You misunderstand. **Prolog-side** jrefs are central to JPL, a "cool feature" indeed:
> 
> ```
> ?- jpl_new('java.util.Date', [], JRef), jpl_call(JRef, toString, [], S).
> JRef = <jref>(00000137ECA1E840),
> S = 'Sat Oct 13 12:52:36 BST 2018'.
> ```

phew...

> (I maintain production code which depends on them and pays my salary) but I question whether we

production code? 

> need a dedicated Java-side concrete syntax for them. Do you have a code snippet doing something reasonable which cannot be done without using org.jpl7.JRef? I think this would have to be Java code which incestuously creates Prolog goals which call back into Java ;-) and maybe this should instead be supported by extending org.jpl7.Atom to represent blobs of type jref.

Yes, suppose an agent has an object `X` (e.g., legacy or third-party infrastructure) that contains useful data (e.g., name of agents, current location of team agents, etc). I want to give Prolog `X` so it can query data from `X`, like `X.location(agent4)`, etc...

It seems to me that  the ability to build a `JRef` on the Java side for an object is one of the coolest features; albeit one of the most sophisticated and hence tricky to use. I do not expect applications will rely on that a lot, but few could be very meaningful to achieve a powerful "glue"

is that what you meant?

**Comment by ssardina** [comment]
Addressed in 2583e9edd74197f63399ececcbf49b9adfe76ac6

In 7.4.x, JVM are either `JPL.JNULL` or `@(null)` Compound. Done and documented.

**Comment by ssardina** [comment]
Addressed in 2583e9edd74197f63399ececcbf49b9adfe76ac6

In 7.4.x, JVM are either `JPL.JNULL` or `@(null)` Compound. Done and documented.
